scipy/stats/_hypotests.py

```diff
@@ -663,3 +664,318 @@ def somersd(x, y=None):
+
+    if pooled:
+        p = (x1 + x2) / (total_col_1 + total_col_2)
+        variances = p * (1 - p) * (1 / total_col_1 + 1 / total_col_2)
+    else:
+        variances = p1 * (1 - p1) / total_col_1 + p2 * (1 - p2) / total_col_2
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide="ignore", invalid="ignore"):
+        wald_statistic = np.divide((p1 - p2), np.sqrt(variances))
+
+    wald_statistic[p1 == p2] = 0  # Removing NaN values
+
+    wald_stat_obs = wald_statistic[table[0, 0], table[0, 1]]
+
+    if alternative == "two-sided":
+        index_arr = np.abs(wald_statistic) >= abs(wald_stat_obs)
+    elif alternative == "less":
+        index_arr = wald_statistic <= wald_stat_obs
+    elif alternative == "greater":
+        index_arr = wald_statistic >= wald_stat_obs
+    else:
+        msg = (
+            "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+            f" found {alternative!r}"
+        )
+        raise ValueError(msg)
+
+    p_value = _binomial_maximisation_of_p_value_with_nuisance_param(
+        total_col_1, total_col_2, index_arr, n_iter
+    )
+    return BarnardExactResult(wald_stat_obs, p_value)
+
+
+def _binomial_maximisation_of_p_value_with_nuisance_param(
+    total_c1, total_c2, idx, n_iter
+):
+    r"""
+    Maximisation of the pvalue in respect of a nuisance parameter considering
+    a 2x2 sample space.
+
+    Barnard exact test iterate over a nuisance parameter
+    :math:`\pi \in [0, 1]` to find the maximum p-value. To search this
+    maxima, I am using a grid search algorithm, reducing :math:`\pi`
+    bounds after each
+    iterations `n_iter`. To compute the p-value, I am using numpy ndarrays
+    which offer powerful broadcast operations. The different nuisance
+    parameters are stored in an ndarray of length 100 and of shape (1, 1,
+    100). Also, to compute the different combination used in the
+    p-values' computation formula, uses `gammaln` which is
+    more tolerant for large value than `scipy.special.comb`. This gives
+    a log combination. For the little precision lost, we gain a lot of
+    performance.
+    """
+    n = total_c1 + total_c2
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1, 1)
+
+    x1_log_comb = _compute_log_combinations(total_c1)
+    x2_log_comb = _compute_log_combinations(total_c2)
+    x1_sum_x2_log_comb = x1_log_comb[x1] + x2_log_comb[x2]
+
+    def get_pvalue_from(nuisance_param):
+
+        with np.errstate(divide="ignore", invalid="ignore"):
+            log_nuisance_arr = np.log(
+                nuisance_param,
+                out=np.zeros_like(nuisance_param),
+                where=nuisance_param >= 0,
+            )
+            log_1_minus_nuisance_arr = np.log(
+                1 - nuisance_param,
+                out=np.zeros_like(nuisance_param),
+                where=1 - nuisance_param >= 0,
+            )
```

Comment:
We don't have arrays anymore. So the parameters and name of the variables have to be adapted.

Reply:
from the doc of `sgho` : `where x is the argument in the form of a 1-D array`. So `nuisance_param` is still a array (and maybe should be named `nuisance_arr` as before).

Reply:
Changes are in f4166d0a43219fa3d3a9e64f9e9aae895b197529. I decided to stick with *arr* suffixe because of @jb-leger comment

Reply:
I edited my comments, formally it is a array. But a 1d array of size 1, so… I do not have opinion.

Reply:
Yes I finally re-did the old version 😉